### PR TITLE
fix(admin): return 200 immediately on shutdown when no pipelines are …

### LINKED ISSUE #3465: POST /api/v1/groups/shutdown?wait=true times out when observed state store is empty
### Pre-filing checklist
- [x] I searched existing issues and didn't find a duplicate

### Component(s)
admin

### Bug Description

Noticed a minor hang in the admin API when calling `POST /api/v1/groups/shutdown?wait=true` on a fresh engine instance without any active pipelines.

Currently in `crates/admin/src/pipeline_group.rs`, the loop checks if all pipelines have terminated like this:

​```rust
let snapshot = state.observed_state_store.snapshot();
let all_terminated =
    !snapshot.is_empty() && snapshot.values().all(|status| status.is_terminated());
​```

Because of `!snapshot.is_empty()`, an empty snapshot makes `all_terminated` false, so the handler polls until `timeout_secs` (default 60s) and returns a `504`.

Fix:

​```rust
let all_terminated =
    snapshot.is_empty() || snapshot.values().all(|status| status.is_terminated());
​```

Happy to send a PR + test for this.

### Steps to Reproduce

1. Start admin server with no pipelines configured.
2. `curl -X POST "http://localhost:8080/api/v1/groups/shutdown?wait=true"`
3. Request hangs for 60s, then fails with HTTP 504.

### Expected Behavior

Empty snapshot → shutdown returns `200 OK` immediately.

### Actual Behavior

Times out after 60s with `504 Gateway Timeout`.

### OTel-Arrow Version

main

### Environment

- OS: Windows / Linux
- Rust: 1.85+

### Configuration

​```yaml
default config, no pipelines
​```

### Log Output

​```
504 Gateway Timeout after 60s
​```

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by Dipanshusinghh**
Fix: `shutdown_all_pipelines` unnecessarily waits full timeout when no pipelines are registered

Fixes #3465

### Problem

`shutdown_all_pipelines` determines whether all pipelines have terminated using:

​```rust
let all_terminated =
    !snapshot.is_empty() && snapshot.values().all(|status| status.is_terminated());
​```

The `!snapshot.is_empty()` guard means that when the snapshot is empty (i.e. no pipelines have been registered yet), `all_terminated` always evaluates to `false` — even though there's technically nothing left to terminate. As a result, the handler stays stuck in the polling loop for the entire `timeout_secs` window (60s by default) before finally returning a `504`.

### Fix

​```rust
let all_terminated =
    snapshot.is_empty() || snapshot.values().all(|status| status.is_terminated());
​```

If the snapshot is empty, there's simply nothing to wait for — so we should short-circuit and return `200` immediately instead of burning the full timeout.

### Testing

Added a regression test covering the empty-snapshot case to make sure this doesn't regress silently in the future.

### CODE REVIEW

File: rust/otap-dataflow/crates/admin/src/pipeline_group.rs [hunk 1]
```diff
@@ -638,4 +639,31 @@ mod tests {
             Some("group delete failed")
         );
     }
+
+    /// Scenario: shutdown is requested with `wait=true` on a fresh engine
+    /// that has no pipelines registered (empty observed state store).
+    /// Guarantees: the handler treats an empty snapshot as all-terminated and
+    /// returns HTTP 200 immediately instead of polling until the timeout.
+    #[tokio::test]
+    async fn shutdown_returns_ok_immediately_when_snapshot_is_empty() {
```
Comment (gouslu) [inline rust/otap-dataflow/crates/admin/src/pipeline_group.rs:658]: Can we add regression tests such that if meaning of "all_terminated" changes in the future, it is caught?

Comment (Dipanshusinghh) [reply rust/otap-dataflow/crates/admin/src/pipeline_group.rs:658]: @gouslu hey! the test right here actually is the regression test specifically for the empty snapshot edge case. did u mean adding more tests for the non-empty cases too? (e.g. verifying it still times out if a pipeline is running, and returns 200 if all registered ones are terminated). if so, i can definitely add those right away!

File: rust/otap-dataflow/crates/admin/src/pipeline_group.rs [hunk 2]
```diff
@@ -213,10 +213,11 @@ async fn shutdown_all_pipelines(
             );
         }
 
-        // Check if all pipelines have terminated
+        // Check if all pipelines have terminated. An empty snapshot means no
+        // pipelines were ever registered, which is itself a terminal state.
         let snapshot = state.observed_state_store.snapshot();
         let all_terminated =
-            !snapshot.is_empty() && snapshot.values().all(|status| status.is_terminated());
+            snapshot.is_empty() || snapshot.values().all(|status| status.is_terminated());
```
Comment (lalitb) [inline rust/otap-dataflow/crates/admin/src/pipeline_group.rs]: Does an empty snapshot always mean there are no running pipelines? The controller and observed-state store are separate, so the snapshot may still be empty while a pipeline is running. In that case, this would return 200 before shutdown finishes. Could we check the controller’s actual runtime state, or add a test for a running pipeline with an empty snapshot?

Comment (Dipanshusinghh) [reply rust/otap-dataflow/crates/admin/src/pipeline_group.rs]: i missed that race condition between the controller and the store.  i will look into checking the controller's actual runtime state instead and push a fix

Comment (lalitb) [reply rust/otap-dataflow/crates/admin/src/pipeline_group.rs]: Thanks for addressing this. However, `engine_config_snapshot()` is still committed configuration, not the controller's actual runtime state. It also does not reliably account for the system observability runtime. So an empty observed snapshot can still return `200` while a runtime is active. Could we check actual shutdown completion from the controller instead, and add the active-runtime/empty-snapshot test mentioned above?

**Comment by Dipanshusinghh** [comment]
@lalitb PtAL! 
noticed the codecov drop should I add the non-empty snapshot tests to fix it r is this approach good?

File: rust/otap-dataflow/crates/admin/src/lib.rs [hunk 3]
```diff
@@ -100,6 +100,11 @@ pub trait ControlPlane: Send + Sync {
     /// Requests shutdown of all currently running runtime instances.
     fn shutdown_all(&self, timeout_secs: u64) -> Result<(), ControlPlaneError>;
 
+    /// Returns true if there are active runtime instances still running.
```
Comment (gouslu) [inline rust/otap-dataflow/crates/admin/src/lib.rs:119]: comment says returns true, but it always returns false.

File: rust/otap-dataflow/crates/controller/src/live_control/mod.rs [hunk 4]
```diff
@@ -467,6 +467,15 @@ impl<
         self.runtime.request_shutdown_all(timeout_secs)
     }
 
+    fn has_active_instances(&self) -> bool {
+        let state = self
+            .runtime
+            .state
+            .lock()
+            .unwrap_or_else(|poisoned| poisoned.into_inner());
+        state.active_instances > 0
```
Comment (gouslu) [inline rust/otap-dataflow/crates/controller/src/live_control/mod.rs]: It looks like this can be 0 while instances are still being spooled up, maybe worth documenting that behavior if this is intended.

**Review by gouslu — APPROVED** [review]

File: rust/otap-dataflow/crates/admin/src/pipeline_group.rs (hunk 1 above)
Comment (gouslu) [reply rust/otap-dataflow/crates/admin/src/pipeline_group.rs:658]: Sorry for the late reply! Yes, non-empty cases. I think that if we cover all possible branches of the conditions in the tests, then if there are any changes to the code that alters the outcome in any of the combinations of conditions then we catch it.

File: rust/otap-dataflow/crates/admin/src/pipeline_group.rs [hunk 5]
```diff
@@ -283,6 +290,18 @@ mod tests {
...
+        ) -> Result<otap_df_config::engine::OtelDataflowSpec, ControlPlaneError> {
+            Ok(otap_df_config::engine::OtelDataflowSpec {
+                version: "1.0".to_string(),
+                policies: Default::default(),
+                topics: Default::default(),
+                engine: Default::default(),
+                groups: Default::default(),
+            })
+        }
+
```
Comment (jmacd) [inline rust/otap-dataflow/crates/admin/src/pipeline_group.rs]: Is this dead code?

File: rust/otap-dataflow/crates/controller/src/live_control/mod.rs (hunk 4 above)
Comment (lalitb) [reply rust/otap-dataflow/crates/controller/src/live_control/mod.rs]: Good catch. I think this is a correctness issue, not just something to document.

Could we make the launch registration and shutdown barrier atomic, or include in-progress lifecycle operations in the completion check? Please also add a deterministic test for this race.

**Review by lalitb — CHANGES_REQUESTED** [review]
Thanks for working through this and for adding the runtime-state check. This is moving in the right direction.

I am requesting changes because the startup race noted above [here](https://github.com/open-telemetry/otel-arrow/pull/3490#discussion_r3693861374) still let shutdown return `200 OK` while a pipeline instance continues running. I know this is a subtle lifecycle edge case, and it is easy to miss. Once that race is handled and covered by a deterministic test, I will be happy to re-review.

**Comment by Dipanshusinghh** [comment]
@lalitb all your concerns regarding the startup race condition have been addressed and covered with a deterministic test also fixed the CI/clippy failures. 
PTAL!

File: rust/otap-dataflow/crates/admin/src/pipeline_group.rs (hunk 5 above)
Comment (Dipanshusinghh) [reply rust/otap-dataflow/crates/admin/src/pipeline_group.rs]: yes, it was unused test mock code. It has been removed in the latest commits.
thanks 4 the review !

File: rust/otap-dataflow/crates/controller/src/live_control/mod.rs [hunk 6]
```diff
@@ -467,6 +467,18 @@ impl<
...
+            || !state.runtime_instances.is_empty()
```
Comment (lalitb) [inline rust/otap-dataflow/crates/controller/src/live_control/mod.rs]: Can we clarify whether this should check active runtime threads or all retained runtime records?

`runtime_instances` can contain `Exited` records. They are not pruned while a pipeline operation, including `active_shutdowns`, still owns the pipeline. This means an exited record can keep this method `true` even when no runtime thread is running.

If this method is only about runtime liveness, could we filter on `RuntimeInstanceLifecycle::Active`, as `wait_until_all_producer_instances_exit()` already does? If shutdown operations must also finish, I think we should check that state explicitly instead of depending on an exited record remaining in this map.

File: rust/otap-dataflow/crates/controller/src/live_control_tests.rs [hunk 7]
```diff
@@ -3733,6 +3733,29 @@ fn runtime_thread_panic_populates_error_source_in_observed_status() {
     assert!(source.contains("backtrace:"));
 }
 
+#[tokio::test]
+async fn has_active_instances_checks_multiple_state_fields() {
```
Comment (lalitb) [inline rust/otap-dataflow/crates/controller/src/live_control_tests.rs:5821]: Could we add a deterministic test for the actual startup race?

The current test sets fields directly and checks the same boolean expression used by `has_active_instances()`. It does not cover shutdown arriving after a pipeline thread starts but before `register_launched_instance()` records it.

Please add a controlled test that pauses the launch in that window, requests shutdown, and verifies shutdown cannot report completion. This is the main correctness case the new guard is intended to protect.

**Comment by Dipanshusinghh** [comment]
@lalitb @jmcd i pushed updates for the remaining feedback. added the missing docs and also added a controlled test to pause the pipeline launch right before registration, so we can verify the shutdown doesnt report early completion during that startup race.
also fixed the test timeout issues that were failing ci earlier.
(btw the current weaver-live-check ci failure is just a flaky infra issue, jq gave an EOF parse error on a truncated json report. i triggered a re run for it)
let me know if it looks good now!

File: rust/otap-dataflow/crates/controller/src/live_control_tests.rs (hunk 7 above)
Comment (lalitb) [reply rust/otap-dataflow/crates/controller/src/live_control_tests.rs:5821]: Thanks for adding the controlled launch pause. This still does not request shutdown while the launch is paused, so it does not reproduce the race from this comment.

Could we keep the launch paused, request shutdown concurrently, verify it does not complete early, then release the launch and verify the new instance receives shutdown and shutdown eventually completes?

File: rust/otap-dataflow/crates/controller/src/live_control/mod.rs (hunk 4 above)
Comment (lalitb) [reply rust/otap-dataflow/crates/controller/src/live_control/mod.rs]: Thanks, this is improved, but I think one in-progress window is still missing.

Per-pipeline operations first add `pipeline_operation_reservations`, then later add `active_rollouts` or `active_shutdowns`. During that gap there may be no registered instance and no `active_engine_operation`, so this check can still return `false` and shutdown can return `200` too early.

Could we include `pipeline_operation_reservations` in the completion check? It may be safer to keep one shared "pending lifecycle work" predicate on `ControllerRuntimeState`, including `global_shutdown_coordinators`, so these state checks do not drift.

Comment (lalitb) [reply rust/otap-dataflow/crates/controller/src/live_control/mod.rs]: There is also a second part to this race: `request_shutdown_all()` only sends shutdown to instances already in its snapshot. If an instance registers afterward, `register_launched_instance()` does not check `global_shutdown_requested`, so that instance never receives shutdown and can keep running until the admin request times out.

I think the fix also needs to handle shutdown during registration. The decision should be made while updating controller state, but the shutdown message should be sent after releasing the state lock.

**Comment by lalitb** [comment]
@Dipanshusinghh  -  Thanks for the update. This is closer, but I think the startup/shutdown race still has two gaps:

  1. The completion check does not include `pipeline_operation_reservations`, so a per-pipeline operation can still be missed before it appears in `active_rollouts` or `active_shutdowns`.
  2. An instance registered after `request_shutdown_all()` takes its snapshot does not receive the shutdown request.

The new test pauses the launch, but it does not request shutdown while the launch is paused, so it does not verify either case yet.

I added details in the existing threads. Once those cases are handled and covered by the test, I will re-review.

File: rust/otap-dataflow/crates/controller/src/live_control/runtime.rs [hunk 8]
```diff
@@ -158,9 +166,24 @@ impl<
...
+                None
+            };
+
+            (should_compact, pending_exit, shutdown_sender)
         };
 
+        if let Some(sender) = shutdown_sender {
+            // Lalit: "the shutdown message should be sent after releasing the state lock."
+            let _ = sender.try_send_shutdown(
+                Instant::now() + Duration::from_secs(60),
```
Comment (lquerel) [inline rust/otap-dataflow/crates/controller/src/live_control/runtime.rs]: [P1] Preserve the requested shutdown deadline for late registrations

request_shutdown_all derives a deadline from timeout_secs, but the state records only a boolean. Any launch that finishes after that request receives a fresh 60-second deadline here, so it can remain active beyond the caller timeout and delay the shutdown coordinator or a later engine operation. Store the global shutdown deadline and reason in state, then apply its remaining time to late registrations instead of starting a new budget.

File: rust/otap-dataflow/crates/controller/src/live_control_tests.rs [hunk 9]
```diff
@@ -5790,6 +5790,110 @@ fn runtime_thread_panic_populates_error_source_in_observed_status() {
     assert!(source.contains("backtrace:"));
 }
 
+static LAUNCH_PAUSE: Mutex<Option<(std::sync::mpsc::Sender<()>, std::sync::mpsc::Receiver<()>)>> =
+    Mutex::new(None);
+
+fn blocking_create(
+    _pipeline_ctx: PipelineContext,
+    node: otap_df_engine::node::NodeId,
```
Comment (lquerel) [inline rust/otap-dataflow/crates/controller/src/live_control_tests.rs]: [P1] Update the post-rename crate paths so the controller tests compile

This new helper still refers to otap_df_engine and otap_df_config, which no longer exist after the crate rename merged from main. Required builds fail with E0433 at these signatures before the new race test can run. Use the current otel_arrow_dfe_engine and otel_arrow_dfe_config paths, or the already imported types, throughout this helper.

File: rust/otap-dataflow/crates/admin/src/pipeline_group.rs [hunk 10]
```diff
@@ -641,4 +649,197 @@ mod tests {
...
+        let app_state = AppState {
+            observed_state_store: store.handle(),
+            metrics_registry,
+            controller: stub(
+                Ok(None),
+                Ok(PipelineGroupConfig::new()),
+                Ok(delete_status("succeeded")),
+            ),
+            terminal_control_plane_permits: Arc::new(tokio::sync::Semaphore::new(1)),
+            heap_profile_permits: Arc::new(tokio::sync::Semaphore::new(1)),
```
Comment (lquerel) [inline rust/otap-dataflow/crates/admin/src/pipeline_group.rs]: [P1] Populate the new CPU-profile semaphore in both AppState values

Current main added the required cpu_profile_permits field to AppState, but both literals introduced by these shutdown tests stop at heap_profile_permits. The admin test target will fail with E0063 once the earlier controller errors are cleared. Add cpu_profile_permits to both new literals so the tests compile against the merged base.

File: rust/otap-dataflow/crates/admin/src/pipeline_group.rs [hunk 11]
```diff
@@ -641,4 +649,197 @@ mod tests {
...
+        .into_response();
+
+        assert_eq!(
+            response.status(),
+            StatusCode::GATEWAY_TIMEOUT,
+            "should time out because an active runtime is still running"
+        );
+    }
+
+    #[tokio::test]
```
Comment (lquerel) [inline rust/otap-dataflow/crates/admin/src/pipeline_group.rs]: [P2] Document the contracts of both new shutdown tests

This repository requires every test to have immediately preceding Scenario and Guarantees comments. The tests starting here and at shutdown_times_out_when_snapshot_is_non_empty_but_not_all_terminated have neither, so their regression contracts are not reviewable without reading their implementations. Add specific two-line contracts above both declarations.

File: rust/otap-dataflow/crates/controller/src/live_control_tests.rs (hunk 7 above)
Comment (Dipanshusinghh) [reply rust/otap-dataflow/crates/controller/src/live_control_tests.rs:5821]: fixed

**Comment by Dipanshusinghh** [comment]
@lalitb @jmacd 
Ptal ! whenever u get a chnce

File: rust/otap-dataflow/crates/controller/src/live_control/runtime.rs [hunk 12]
```diff
@@ -141,15 +141,23 @@ impl<
...
         let context_bindings = Arc::clone(&launched.context_bindings);
-        let (should_compact, pending_exit) = {
+        let (should_compact, pending_exit, shutdown_sender) = {
             let mut state = self
                 .state
                 .lock()
                 .unwrap_or_else(|poisoned| poisoned.into_inner());
+
+            let should_shutdown = state.global_shutdown_requested;
+            let control_sender = if should_shutdown {
```
Comment (lalitb) [inline rust/otap-dataflow/crates/controller/src/live_control/runtime.rs:151]: This issue is coming from the latest changes to handle late registrations. The new code clears the retained sender before confirming shutdown delivery. If `try_send_shutdown` fails because the control channel is full, the error is ignored and later shutdown requests have no sender to retry.

Could we retain the sender until delivery succeeds and report failures, as the normal shutdown path does? Please also cover a failed send followed by a successful retry.

File: rust/otap-dataflow/crates/controller/src/live_control/runtime.rs [hunk 13]
```diff
@@ -162,9 +170,24 @@ impl<
...
+            } else {
+                None
+            };
+
+            (should_compact, pending_exit, shutdown_sender)
         };
 
+        if let Some(sender) = shutdown_sender {
+            // Send shutdown after releasing the state lock to avoid lock contention or deadlocks.
+            let _ = sender.try_send_shutdown(
```
Comment (lalitb) [inline rust/otap-dataflow/crates/controller/src/live_control/runtime.rs:185]: This issue seems to be coming from the latest changes to handle late registrations. The new shutdown call gives them a fresh 60-second deadline, regardless of the timeout supplied to `shutdown_all`. That can extend a short shutdown request or cut short a longer drain budget.

Could we preserve the global shutdown deadline and use it here? Please also test with a non-default timeout.

**Comment by Dipanshusinghh** [comment]
@lalitb  @jmacd  PTAL  whenever u get a chnce

File: rust/otap-dataflow/crates/controller/src/live_control/runtime.rs (hunk 12 above)
Comment (Dipanshusinghh) [reply rust/otap-dataflow/crates/controller/src/live_control/runtime.rs]: fixed

File: rust/otap-dataflow/crates/controller/src/live_control/runtime.rs (hunk 13 above)
Comment (Dipanshusinghh) [reply rust/otap-dataflow/crates/controller/src/live_control/runtime.rs]: fixed

**Comment by lalitb** [comment]
Thanks @Dipanshusinghh -  The sender and deadline fixes address my earlier comments. One issue remains: shutdown waits only for pipelines in its original list. A pipeline registered later receives shutdown, but observability may stop before it finishes, losing its final telemetry.

Could we also wait for pipelines still starting or registered later? Please add a test that pauses just before controller registration and checks that observability stays running until the late pipeline exits. Pausing receiver creation does not guarantee registration is paused.
